aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -0,0 +1,2431 @@
+        if (amount == 0) { return };
+
+        // fail unlock of more stake than `active` on the stake pool
+        let (active, _, _, _) = stake::get_stake(pool_address);
+        assert!(amount <= active, error::invalid_argument(ENOT_ENOUGH_ACTIVE_STAKE_TO_UNLOCK));
+
+        // synchronize delegation and stake pools before any user operation
+        synchronize_delegation_pool(pool_address);
+
+        let pool = borrow_global_mut<DelegationPool>(pool_address);
+        let delegator_address = signer::address_of(delegator);
+
+        amount = redeem_active_shares(pool, delegator_address, amount);
+
+        stake::unlock(&retrieve_stake_pool_owner(pool), amount);
+        buy_in_pending_inactive_shares(pool, delegator_address, amount);
+
+        event::emit_event(
+            &mut pool.unlock_stake_events,
+            UnlockStakeEvent {
+                pool_address,
+                delegator_address,
+                amount_unlocked: amount,
+            },
+        );
+    }
+
+    /// Move `amount` of coins from pending_inactive to active.
+    public entry fun reactivate_stake(delegator: &signer, pool_address: address, amount: u64) acquires DelegationPool {
+        // short-circuit if amount to reactivate is 0 so no event is emitted
+        if (amount == 0) { return };
+        // synchronize delegation and stake pools before any user operation
+        synchronize_delegation_pool(pool_address);
+
+        let pool = borrow_global_mut<DelegationPool>(pool_address);
+        let delegator_address = signer::address_of(delegator);
+
+        let observed_lockup_cycle = pool.observed_lockup_cycle;
+        amount = redeem_inactive_shares(pool, delegator_address, amount, observed_lockup_cycle);
+
+        stake::reactivate_stake(&retrieve_stake_pool_owner(pool), amount);
+        pool_u64::buy_in(&mut pool.active_shares, delegator_address, amount);
+
+        event::emit_event(
+            &mut pool.reactivate_stake_events,
+            ReactivateStakeEvent {
+                pool_address,
+                delegator_address,
+                amount_reactivated: amount,
+            },
+        );
+    }
+
+    /// Withdraw `amount` of owned inactive stake from the delegation pool at `pool_address`.
+    public entry fun withdraw(delegator: &signer, pool_address: address, amount: u64) acquires DelegationPool {
+        // short-circuit if amount to withdraw is 0 so no event is emitted
+        if (amount == 0) { return };
+        // synchronize delegation and stake pools before any user operation
+        synchronize_delegation_pool(pool_address);
+        withdraw_internal(borrow_global_mut<DelegationPool>(pool_address), signer::address_of(delegator), amount);
+    }
+
+    fun withdraw_internal(pool: &mut DelegationPool, delegator_address: address, amount: u64) {
+        // short-circuit if amount to withdraw is 0 so no event is emitted
+        if (amount == 0) { return };
```

Comment:
After some testing, I think it might be cleaner if we emit an error here when there's withdraw amount is 0. This would be less confusing for the users as the tx still goes through but they get nothing

Reply:
I think we should assert only on the external withdraw, the internal one is called also when automatically executing the inactive pending withdrawal. The invariant is that if the withdrawal exists then its stake != 0, and would never assert on this line. Supposing this is not true, when distributing pending_inactive commission rewards, the withdrawal would be executed and assert, halting all ops on the pool.

Reply:
Makes sense